### PR TITLE
Add `logAeson` (#578)

### PR DESCRIPTION
**Pull request by prutz1311**
Starts work on  #578  .

### Pre-review checklist

- [x] All code has been formatted using our config (`make format` for Purescript, `nixpkgs-fmt` for Nix)
- [x] All Purescript imports are explicit, including constructors
- [ ] **All** existing examples have been run locally against a fully-synced testnet node
- [x] The integration and unit tests have been run (`npm run test`) locally
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`)

### CODE REVIEW

**Review by ngua — COMMENTED** [review]
I think it would be nice to include partially-applied `logAeson` variants for users as well: `logAesonInfo`, `logAesonDebug`, etc...

**Review by bladyjoker — DISMISSED** [review]

**Review by vvtran — COMMENTED** [review]
I think other than what @ngua mentioned and conflicts, this looks fine.

**Comment by ngua** [comment]
I'm going to mark this as a draft again since it needs some more work

File: src/Types/RedeemerTag.purs [hunk 1]
```diff
@@ -37,3 +40,12 @@ instance BoundedEnum RedeemerTag where
...
   fromEnum = genericFromEnum
+
+instance EncodeAeson RedeemerTag where
+  encodeAeson' = case _ of
+    Spend -> encodeTagged' "Spend" {}
+    Mint -> encodeTagged' "Mint" {}
+    Cert -> encodeTagged' "Cert" {}
+    Reward -> encodeTagged' "Reward" {}
+    where
+    encodeTagged' str x = encodeAeson' $ encodeTagged str x (Op encodeAeson)
```
Comment (ngua) [inline src/Types/RedeemerTag.purs:51]: Missing a type signature here

Comment (rmgaray) [reply src/Types/RedeemerTag.purs:51]: Thanks, this should have been replaced by the utility I introduced in the `Helpers` module

**Review by vvtran — COMMENTED** [review]
Some comments

File: src/Types/TransactionMetadata.purs [hunk 2]
```diff
@@ -64,3 +69,12 @@ derive instance Generic TransactionMetadatum _
 
 instance Show TransactionMetadatum where
   show x = genericShow x
+
+instance EncodeAeson TransactionMetadatum where
+  encodeAeson' = case _ of
+    MetadataMap m -> encodeAeson' $ encodeMap m
```
Comment (klntsky) [inline src/Types/TransactionMetadata.purs:75]: Why not `encodeTagged' "Map"`?

File: src/Types/RedeemerTag.purs [hunk 3]
```diff
@@ -26,3 +36,23 @@ derive instance Ord RedeemerTag
...
+
+instance Enum RedeemerTag where
+  pred = genericPred
+  succ = genericSucc
+
+instance Bounded RedeemerTag where
+  top = Reward
+  bottom = Spend
+
+instance BoundedEnum RedeemerTag where
```
Comment (klntsky) [inline src/Types/RedeemerTag.purs:48]: I don't see how can this instance be used. I think the order of the constructors is rather arbitrary. Please provide a justification or remove (together with `Bounded` and `Enum`).

File: src/Contract/Monad.purs [hunk 4]
```diff
@@ -370,3 +381,60 @@ logWarn' = Logger.warn Map.empty
 logError'
   :: forall (m :: Type -> Type). MonadLogger m => String -> m Unit
 logError' = Logger.error Map.empty
+
+-- Log JSON representation of a data structure
```
Comment (klntsky) [inline src/Contract/Monad.purs:385]: Please use doc-comments

```suggestion
-- | Log JSON representation of a data structure
```

**Review by klntsky — CHANGES_REQUESTED** [review]

File: src/Types/TransactionMetadata.purs (hunk 2 above)
Comment (rmgaray) [reply src/Types/TransactionMetadata.purs:75]: I forgot to add the tag apparently, thanks.

File: src/Types/RedeemerTag.purs (hunk 3 above)
Comment (rmgaray) [reply src/Types/RedeemerTag.purs:48]: I didn't write this code so I don't really know why it was added. It's part of this commit: [b5d6c32e2](https://github.com/Plutonomicon/cardano-transaction-lib/commit/b5d6c32e2f5b7c6eed1002908aacfc84654d1acc#diff-4393a67a738272abd3fe0cadecc576ddf83d873a34fa56638c79c9e65eb97676)

**Review by klntsky — APPROVED** [review]

**Review by ngua — APPROVED** [review]

**Pull request merged**